### DISCUSSION TITLE
make tgbot-cpp to listen on next message that user sends and handle it

### DISCUSSION THREAD

**Discussion by miannoodle01** [reo7sp/tgbot-cpp#269] [General]
hello! i wish to know how should i wrote a code that for example, on writing "request", bot write this message and listen on the message the text the user will send. for example:
user: request.
bot: write your request.
user: i want the source code of the bot.
bot: your message will be delivered to admins.

**Comment by reo7sp** [comment]
Register the handlers once and keep a per-chat conversation state. Do not add a new `onAnyMessage` handler every time the user starts a request, because those handlers accumulate.

```cpp
std::unordered_set<std::int64_t> waitingForRequest;

bot.getEvents().onCommand("request", [&](TgBot::Message::Ptr message) {
    waitingForRequest.insert(message->chat->id);
    bot.getApi().sendMessage(message->chat->id, "Write your request");
});

bot.getEvents().onNonCommandMessage([&](TgBot::Message::Ptr message) {
    if (!waitingForRequest.erase(message->chat->id)) {
        return;
    }

    const auto request = message->text.value_or("");
    // Save or forward request here.
    bot.getApi().sendMessage(message->chat->id, "Your message will be delivered to admins");
});
```

Store this state in a database if it must survive restarts, and synchronize access when using concurrent webhook handlers.
